File: app/src/main/java/com/vocaby/app/adapters/CustomEntryAdapter.java
package com.vocaby.app.adapters;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.vocaby.app.R;

import java.util.ArrayList;
import java.util.List;

public class CustomEntryAdapter extends RecyclerView.Adapter<CustomEntryAdapter.EntryViewHolder> {
    private List<String> customEntries;
    private final ItemTouchListener itemTouchListener;
    private final MaterialAlertDialogBuilder builder;

    public interface ItemTouchListener {
        void onItemDelete(String entry, int position);
        void onItemTouch(String entry, int position);
    }

    public CustomEntryAdapter(Activity activity, ItemTouchListener itemTouchListener) {
        customEntries = new ArrayList<>();
        this.itemTouchListener = itemTouchListener;
        builder = new MaterialAlertDialogBuilder(activity);
    }

    @NonNull
    @Override
    public EntryViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.custom_entry_item, parent, false);
        return new EntryViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull EntryViewHolder holder, int position) {
        String entry = customEntries.get(holder.getAdapterPosition());

        holder.itemView.setOnClickListener(view -> {
            // You want to use getAdapterPosition instead of position
            // in the listener because onBindViewHolder will not be called again
            // from structural changes.
            itemTouchListener.onItemTouch(
                    entry,
                    holder.getAdapterPosition()
            );
        });

        holder.removeEntryButton.setOnClickListener(v -> {
            builder.setTitle("Are you sure you want to delete?")
                    .setMessage(entry)
                    .setPositiveButton("DELETE", (dialog, which) ->
                            itemTouchListener.onItemDelete(entry, holder.getAdapterPosition()))
                    .setNegativeButton("CANCEL", null);

            showAlertDialog();
        });

        holder.header.setText(customEntries.get(position));
    }

    private void showAlertDialog() {
        AlertDialog alert = builder.create();
        alert.show();
    }

    // The list is only set on create
    @SuppressLint("NotifyDataSetChanged")
    public void setList(List<String> newList) {
        customEntries = newList;
        notifyDataSetChanged();
    }

    public void addEntry() {
        notifyItemInserted(0);
    }

    public void deleteEntry(int position) {
        notifyItemRemoved(position);
    }

    @Override
    public int getItemCount() {
        return customEntries.size();
    }

    public static class EntryViewHolder extends RecyclerView.ViewHolder {
        TextView header;
        ImageButton removeEntryButton;
        public EntryViewHolder(@NonNull View itemView) {
            super(itemView);
            header = itemView.findViewById(R.id.custom_entry_item_header);
            removeEntryButton = itemView.findViewById(R.id.delete_button);
        }
    }

}

File: app/src/main/java/com/vocaby/app/adapters/CustomEntryAdapter.kt
package com.vocaby.app.adapters

import android.annotation.SuppressLint
import android.app.Activity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.vocaby.app.R
import com.vocaby.app.models.customentry.UserEntry
import com.vocaby.app.utils.Formatter

class CustomEntryAdapter(activity: Activity, private val interaction: Interaction) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>(), ItemTouchHelperAdapter {
    private var customEntries: List<UserEntry> = ArrayList()
    private val alertDialogBuilder: MaterialAlertDialogBuilder = MaterialAlertDialogBuilder(activity)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return CustomEntryViewHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.custom_entry_item,
                parent,
                false
            ),
            interaction,
            alertDialogBuilder
        )
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder) {
            is CustomEntryViewHolder -> holder.bind(customEntries[position])
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    fun submitList(list: List<UserEntry>) {
        customEntries = list
        notifyDataSetChanged()
    }

    fun addEntry() {
        notifyItemInserted(0)
    }

    fun deleteEntry(position: Int) {
        notifyItemRemoved(position)
    }

    fun updateEntry(position: Int) {
        notifyItemChanged(position)
    }

    override fun getItemCount(): Int {
        return customEntries.size
    }

    class CustomEntryViewHolder
    constructor(
        itemView: View,
        private val interaction: Interaction,
        private val alertDialogBuilder: MaterialAlertDialogBuilder
    ): RecyclerView.ViewHolder(itemView) {
        private val header: TextView = itemView.findViewById(R.id.custom_entry_item_header)
        private val lastUpdated: TextView = itemView.findViewById(R.id.custom_entry_item_updated)
        private val deleteButton: ImageButton = itemView.findViewById(R.id.delete_button)

        fun bind(userEntry: UserEntry) {
            header.text = userEntry.entry
            lastUpdated.text = Formatter.formatDateStringForDisplay(userEntry.lastUpdated)

            itemView.setOnClickListener {
                interaction.onItemTouch(userEntry.entry, adapterPosition)
            }

            deleteButton.setOnClickListener {
                alertDialogBuilder
                    .setTitle("Are you sure you want to delete?")
                    .setMessage(userEntry.entry)
                    .setPositiveButton("DELETE") { _, _ ->
                        interaction.onItemDelete(userEntry.entry, adapterPosition)
                    }.setNegativeButton("CANCEL", null).create().show()
            }
        }
    }

    interface Interaction {
        fun onItemTouch(entry: String, position: Int)
        fun onItemDelete(entry: String, position: Int)
    }

    override fun onItemMove(fromPosition: Int, toPosition: Int): Boolean {
        return false
    }

    override fun onItemDismiss(position: Int) {
        alertDialogBuilder
            .setTitle("Are you sure you want to delete?")
            .setMessage(customEntries[position].entry)
            .setPositiveButton("DELETE") { _, _ ->
                interaction.onItemDelete(customEntries[position].entry, position)
            }.setNegativeButton("CANCEL") { _, _ -> notifyItemChanged(position) }.create().show()
    }
}